text/0000-grand-testing-unification.md

```diff
@@ -0,0 +1,551 @@
+
+Consider this snippet using existing acceptance test helpers:
+
+```js
+fillIn('.display-title', 'Hello!');
+click('.update-display-title');
+andThen(() => {
+  assert.equal(Ember.$('.display-title').text(), 'Hello!');
+});
+```
+
+There is no indication when reading this snippet that `click` and `fillIn` are async (the reader just has to know the "magical" async test helper names).
+
+I propose the following replacement snippet:
+
+```js
+await this.fillIn('.display-title', 'Hello!');
+await this.click('.update-display-title');
+
+assert.equal(this.find('.display-title').text(), 'Hello!');
+```
+
+The async/await proposal is currently in Stage 3, and I believe that it has been stable enough to rely on by default.
+
+For more details on `async` / `await` see:
+
+- https://tc39.github.io/ecmascript-asyncawait/
+- https://jakearchibald.com/2014/es7-async-functions/
+
+### Shared Test Helpers
+
+By sharing the same test helpers amongst all types of tests (unit, integration, and acceptance) we will have a single consistent mechanism across the board.
+
+#### Always Available
+
+These helpers will be available from all test contexts:
+
+- `pauseTest` - Helper used to pause the test runner and allow interaction.
+- `owner` - Property used to return the public owner instance in use by the test.
+- `registerWaiter` - Used to instruct the system about unknown asynchrony that should be "waited" for throughout the system (i.e. after an `await this.click(..)` we should wait until AJAX requests are completed).
+- `registerHelper` - Used to register a helper with the test context.
+- `registerHelpers` - Used to register multiple helpers with the test context.
+
+#### DOM Helpers
+
+These helpers will be available in all testing contexts where DOM helpers are applicable (acceptance and component integration tests). All DOM related helpers will use normal DOM API's, and will avoid using jQuery unless required.
+
+- `click` - Async helper to click on the specified selector.
+- `triggerKeyEvent` - Async helper to trigger a `KeyEvent` on the specified selector.
+- `triggerEvent` - Async helper to trigger an event on the specified selector.
+- `fillIn` - Async helper to enter text on the specified selector.
+- `find` - Sync helper to return an element for a given selector (via `document.querySelector`).
+- `findAll` - Sync helper to return a list of elements matching the given selector (via `document.querySelectorAll`).
+- `element` - Property to access the raw element at the root of the DOM for the given test.
+
+#### Rendering/Integration Helpers
+
+These helpers will be available in testing contexts that allow rendering individual template snippets (integration test):
+
+- `render`  - Async helper to render a given handlebars template.
+- `clearRender` - Async helper to clear a previously rendered template.
```

Comment:
@martndemus - Yes, but I would suggest that you should use a different test instead of rendering twice most of the time...

Also, @mmun is absolutely correct, each call to `this.render(....)` should call `this.clearRender()` if it had been previously rendered.